### ISSUE TITLE
@sanity/client fails type checking when building a NextJS project

### ISSUE THREAD

**Issue by mstaicu** [sanity-io/sanity#2508]
**Describe the bug**

```
> next build

info  - Loaded env from /Users/mircea/Documents/github/radu-website-nextjs/web/.env.local
info  - Using webpack 5. Reason: no custom webpack configuration in next.config.js https://nextjs.org/docs/messages/webpack5
Failed to compile.

./src/pages/profile.tsx:11:23
Type error: Argument of type '{ projectId: string | undefined; dataset: string | undefined; apiVersion: string; token: string | undefined; useCdn: false; }' is not assignable to parameter of type 'ClientConfig'.
  Type '{ projectId: string | undefined; dataset: string | undefined; apiVersion: string; token: string | undefined; useCdn: false; }' is not assignable to type 'ProjectlessClientConfig'.
    Property 'useProjectHostname' is missing in type '{ projectId: string | undefined; dataset: string | undefined; apiVersion: string; token: string | undefined; useCdn: false; }' but required in type '{ useProjectHostname: false; projectId?: string | undefined; dataset?: string | undefined; }'.

> 11 | const client = sanity({
     |                       ^
  12 |   projectId: process.env.SANITY_IO_PROJECT_ID,
  13 |   dataset: process.env.SANITY_IO_DATASET,
  14 |   apiVersion: 'v1',
```

**To Reproduce**

1. Install the `@sanity/client@2.8.0` in a NextJS project
2. Instantiate the client
```
const client = sanity({
  projectId: process.env.SANITY_IO_PROJECT_ID,
  dataset: process.env.SANITY_IO_DATASET,
  apiVersion: 'v1',
  token: process.env.SANITY_IO_TOKEN,
  useCdn: false,
});
```
3. Build the NextJS application

It seems that the `useProjectHostname` is deprecated in the `BaseClientConfig` but still being used in the `ProjectlessClientConfig` and `ProjectClientConfig`

If I do set the `useProjectHostname` to `false` when instantiating the client, the type checking works but the client fails to fetch.
If I do set the `useProjectHostname` to `true` when instantiating the client,  the type checking fails with the following error ( but the client is able to fetch data )

```
> next build

info  - Loaded env from /Users/mircea/Documents/github/radu-website-nextjs/web/.env.local
info  - Using webpack 5. Reason: no custom webpack configuration in next.config.js https://nextjs.org/docs/messages/webpack5
Failed to compile.

./src/pages/index.tsx:13:23
Type error: Argument of type '{ useProjectHostname: true; projectId: string | undefined; dataset: string | undefined; apiVersion: string; token: string | undefined; useCdn: false; }' is not assignable to parameter of type 'ClientConfig'.
  Types of property 'projectId' are incompatible.
    Type 'string | undefined' is not assignable to type 'string'.
      Type 'undefined' is not assignable to type 'string'.
```

**Comment by ricokahler** [comment]
Hi @mstaicu!

Thanks for bringing this up! and you're spot on.

The sanity constructor accepts two versions of the config:

- a `ProjectlessClientConfig` that allows the `projectId` to be optional and 
- a `ProjectClientConfig` that must have the `projectId` defined and as a string.

https://github.com/sanity-io/sanity/blob/67c0d1ac54f2f214279a6d3107ac789e80fa8809/packages/%40sanity/client/sanityClient.d.ts#L525-L537

The `ProjectlessClientConfig` and the somewhat related `useProjectHostname` are for special use cases (that most users shouldn't care about).

TL;DR If the type of `projectId` is `string | undefined` (instead of just `string`) then typescript will think you meant to fill out the `ProjectlessClientConfig` instead of the standard `ProjectClientConfig`.

You can fix this error by asserting that `projectId` is defined by using the `!` or by throwing if that value is falsy prior to running code that will configure the client.

```diff
// Note: either of these changes will work
+ if (!process.env.SANITY_IO_PROJECT_ID) {
+   throw new Error('SANITY_IO_PROJECT_ID was falsy');
+ }

const client = sanity({
  // Note: either of these changes will work
- projectId: process.env.SANITY_IO_PROJECT_ID,
+ projectId: process.env.SANITY_IO_PROJECT_ID!,
  dataset: process.env.SANITY_IO_DATASET,
  apiVersion: 'v1',
  token: process.env.SANITY_IO_TOKEN,
  useCdn: false,
});
```

This experience is definitely not good DX so we might just allow `projectId` to be optional. Using `process.env.SOME_VAR` is a pretty common use case and the client throws at runtime if that value is undefined anyway.

**Comment by mstaicu** [comment]
Oh nice, thanks for that
